capsules/src/virtual_digest.rs

```diff
@@ -0,0 +1,136 @@
+    for VirtualMuxDigest<'a, A, T>
+{
+    /// Set the client instance which will receive `add_data_done()` and
+    /// `hash_done()` callbacks
+    fn set_client(&'a self, client: &'a dyn digest::Client<'a, T>) {
+        self.client.set(client);
+    }
+
+    /// Add data to the digest IP.
+    /// All data passed in is fed to the Digest hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ReturnCode, &'static mut [u8])> {
+        if self.enabled.get() == true {
+            self.mux.digest.add_data(data)
+        } else {
+            Err((ReturnCode::EBUSY, data.take()))
+        }
+    }
+
+    /// Request the hardware block to generate a Digest
+    /// This doesn't return anything, instead the client needs to have
+    /// set a `hash_done` handler.
+    fn run(&'a self, digest: &'static mut T) -> Result<(), (ReturnCode, &'static mut T)> {
+        if self.enabled.get() == true {
+            self.mux.digest.run(digest)
+        } else {
+            Err((ReturnCode::EBUSY, digest))
+        }
+    }
+
+    /// Disable the Digest hardware and clear the keys and any other sensitive
+    /// data
+    fn clear_data(&self) {
+        if self.enabled.get() == true {
+            self.enabled.set(false);
+            self.mux.running.set(false);
+            self.mux.digest.clear_data()
+        }
+    }
+}
+
+impl<A: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> digest::Client<'a, T>
+    for VirtualMuxDigest<'a, A, T>
+{
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]) {
+        self.client
+            .map(move |client| client.add_data_done(result, data));
+    }
+
+    fn hash_done(&'a self, result: Result<(), ReturnCode>, digest: &'static mut T) {
+        self.client
+            .map(move |client| client.hash_done(result, digest));
+    }
+}
+
+impl<A: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> digest::HMACSha256
+    for VirtualMuxDigest<'a, A, T>
+{
+    fn set_mode_hmacsha256(&self, key: &[u8]) -> Result<(), ReturnCode> {
+        // Check if any mux is enabled. If it isn't we enable it for us.
+        if self.mux.running.get() == false {
+            self.mux.running.set(true);
+            self.enabled.set(true);
+            self.mux.digest.set_mode_hmacsha256(key)
+        } else {
+            Err(ReturnCode::EBUSY)
+        }
+    }
+}
+
+/// Calling a 'set_mode*()' function from a `VirtualMuxDigest` will mark that
+/// `VirtualMuxDigest` as the one that has been enabled and running. Until that
+/// Mux calls `clear_data()` it will be the only `VirtualMuxDigest` that can
+/// interact with the underlying device.
+pub struct MuxDigest<'a, A: kernel::hil::digest::HMACSha256 + digest::Digest<'a, T>, T: DigestType>
+{
+    digest: &'a A,
+    running: Cell<bool>,
```

Comment:
```suggestion
    running: OptionalCell<usize>,
```

I think keeping track of which `VirtualMuxDigest` is running is the missing piece to enabling virtualization. Then only that client can use it until the digest engine is no longer running.

Reply:
I couldn't figure out a good way to keep track of which one is running. Any ideas?

Currently we just rely on the one that starts to say that it is running.

Reply:
My first thought would be to assign an "id" when the virtual client is created, by having the mux increment a counter every time a new client is created. Then the mux can use that id to mark who currently owns the digest HW.